Added Zoom Link

abirmahmood6 opened the pull request:
Added the Zoom link feature to Discord Bot.
The discord bot will send the programming club Zoom link in the discord channel every Wednesday at 10 am (10 am is what I set for now but the time can be changed).

abirmahmood6 committed:
sample

abirmahmood6 committed:
edit gitignore

abirmahmood6 committed:
Wednesdays Zoom Link

Managed to send zoom link on every Wednesday.
Still working on how to send the message at a specific time like 10am or so.

abirmahmood6 committed:
Added Specific time

Coded the part of sending message at specific time in the discord channel

p5quared commented on main.py:
```diff
@@ -55,10 +55,27 @@ def write_cache(live_cache: dict):
     with open("roles.json", "w") as f:
         json.dump(live_cache, f, indent=2)
 
+#Below is the code for sending ZOOM LINK by Abir
```
How have you tested this function? It looks like most code will never be reached?

At the  start, we get the day. Let's imagine today is Friday. "now" should be equal to 4.

Therefore...
```python3
if now == 2: (False)
    ...skips everything
```

abirmahmood6 replied:
I thought the zoom link is supposed to be sent every week only in Wednesdays. As the club meetings are only on Wednesdays. Thus, if now = 2 (True), the zoom link will be sent out to the channel on every Wednesday before the meeting.

p5quared commented on main.py:
```diff
@@ -55,10 +55,27 @@ def write_cache(live_cache: dict):
 @bot.event
 async def on_ready():
     print(f'Bot is logged in as {bot.user}')
+    await zoom_link()
```
This function is being called in `on_ready()`, a default function that runs once every time the bot starts up. In other words, this function zoom_link is called once when the bot starts up and never again. 

<s>I think the solution may be to create some sort of loop using dateTime that runs the function once per week. 

Here's a couple resources to look at:
https://pypi.org/project/schedule/
https://youtu.be/ovT9GQ-0mlU</s>

Wait never mind, I did some digging and found this:
https://discordpy.readthedocs.io/en/stable/ext/tasks/index.html?highlight=loop#discord.ext.tasks.Loop

It is a decorator which is like these: `@bot.command()`, `@commands.has_permissions(administrator=True)`, etc...

You can "decorate" your function and then it will automatically run once per week!
Read the documentation on how to do it. I actually found this from someone else's discord bot here:
https://github.com/HunterSandlin/DiscordDailyMessageBot/blob/master/DailyMessageBot.py

abirmahmood6 replied:
Fixed! I made two changes. Firstly, added tasks.loop(hours=24) on top of the zoom_link function. This creates a loop that will iterate the zoom_link function every 24 hours when program is running. 
Another change is - under the on_ready function, I added zoom_link.start(). This is important because this line is what starts the loop. Therefore, now the zoom_link function will automatically run once per week.

p5quared commented on main.py:
```diff
@@ -83,13 +100,17 @@ async def on_message(message):
 
 
 @bot.command()
+@commands.has_permissions(administrator=True)
```
I think this is my fault but these changes which are from my version should not appear as commits in this PR. Likely a problem with how I have the repo structured. Will look into changing this for the next release.

p5quared reviewed (changes requested):
I just wanna say I am very impressed! I believe there are some issued which I've outlined, but you've done a great job jumping right into the async await syntax and working with datetime.

Not sure if you've used datetime in previous projects but it's a cool implementation you've got.

Take a look at my comments and resubmit changes whenever you have time!
p.s. please double check your branch sync
[explained](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork)

abirmahmood6 commented:
I am making a new pull request. This time it will automatically run the function every week when program is running.

abirmahmood6 committed:
Added Loop to Zoom link

I coded the zoom_link function to loop every 24 hour when the program is on running.

abirmahmood6 commented:
I pushed the updated code in the same pull request. Have a look and let me know if need any changes. 
Also some feedback :) @p5quared

p5quared committed:
Merge branch 'main' into welcome
